src/core/include/mp-units/bits/scaling.h

```diff
@@ -0,0 +1,229 @@
+    } else if constexpr (_is_integral(_pow<-1>(M))) {
+      return static_cast<To>(static_cast<value_type_t<From>>(value) / _get_value<U>(den));
+    } else {
+      return static_cast<To>(ratio<U>.scale(static_cast<value_type_t<From>>(value)));
+    }
+  }
+};
+
+
+template<typename Rep>
+struct fixed_point_scaling_traits {
+  template<Magnitude M, typename From>
+  static constexpr Rep scale_from(M, const From& value)
+  {
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  };
+
+  template<Magnitude M>
+  static constexpr auto scale(M, const Rep& value)
+  {
+    // for standard integer types, the result respresentation is always the same
+    return fixed_point_scaling_impl<M{}>::template scale<Rep>(value);
+  }
+};
+
+
+template<typename T, typename Other = T>
+inline constexpr auto select_scaling_traits = [] {
+  if constexpr (requires {
+                  // we only check if the traits class is complete, not it's members; we do not want to fall-back
+                  // depending on the argument types
+                  { sizeof(mp_units::scaling_traits<T>) } -> std::integral;
+                }) {
+    // traits class is defined; use that
+    return mp_units::scaling_traits<T>{};
+  } else {
+    // undefined traits class; fall-back to default handling
+    if constexpr (mp_units::treat_as_floating_point<T> || mp_units::treat_as_floating_point<Other>) {
+      return floating_point_scaling_traits<T>{};
+    } else if constexpr (std::is_integral_v<T> ||
+                         (std::is_integral_v<value_type_t<T>> && std::convertible_to<value_type_t<T>, T> &&
+                          std::convertible_to<T, value_type_t<T>>)) {
+      return fixed_point_scaling_traits<value_type_t<T>>{};
+    } else {
+      // placeholder to report failure
+      return std::false_type{};
+    }
+  }
+}();
+
+template<typename T, typename Other = T>
+concept HasScalingTraits = !std::convertible_to<decltype(select_scaling_traits<T, Other>), std::false_type>;
+
+}  // namespace detail
+
+template<typename To, Magnitude M, typename From>
+  requires detail::HasScalingTraits<To, From> ||
+           requires(const From& value) { value.scale(std::type_identity<To>{}, M{}); }
+constexpr To scale(std::type_identity<To> to_type, M scaling_factor, const From& value)
+{
+  if constexpr (requires {
+                  { value.scale(to_type, scaling_factor) } -> std::convertible_to<To>;
+                }) {
+    return value.scale(to_type, scaling_factor);
+  } else {
+    return detail::select_scaling_traits<To, From>.scale_from(scaling_factor, value);
+  }
+}
+
+template<Magnitude M, typename From>
+  requires detail::HasScalingTraits<From> || requires(const From& value) { value.scale(M{}); }
+constexpr auto scale(M scaling_factor, const From& value)
+{
+  if constexpr (requires { value.scale(scaling_factor); }) {
+    return value.scale(scaling_factor);
```

Comment:
Fine with me. Explicit specialisations are sometimes a bit painful to do because of the namespaces. That said, there are plenty of precedents in the standard library. ADL is a bit more convenient, and modern approaches would go through a CPO. On the other hand, I would actually prefer not to dispatch on the magnitude type at all, because if a different implementation is chosen for different magnitudes by accident, that would be very confusing. Instead, all paths for different magnitudes should remain close and under the control of the same "customiser". Thus, I think I'll give it another shot just explicit specialisation of `mp_units::scaling_traits<From, To=UnspecifiedRepresentation>`.